### PR TITLE
Another version of EMNA (with population size adaptation)

### PR DESCRIPTION
**Pull request by fteytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Adding a robust EMNA algorithm with better performances in a non-parallel context.

## How Has This Been Tested (if it applies)

- Need more test
- Need cleaning

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/optimization/optimizerlib.py [hunk 1]
```diff
@@ -1492,24 +1492,33 @@ def __init__(
             budget: Optional[int] = None,
             num_workers: int = 1,
             isotropic: bool = True,
-            naive: bool = True
+            naive: bool = True,
+            populationSizeAdapatation: bool = False
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: please use standard instance/attribute name: `population_size_adapatation`

File: nevergrad/optimization/optimizerlib.py [hunk 2]
```diff
@@ -1492,24 +1492,33 @@ def __init__(
...
     ) -> None:
         super().__init__(parametrization, budget=budget, num_workers=num_workers)
         self.isotropic: bool = isotropic
         self.naive: bool = naive
+        self.populationSizeAdapatation = populationSizeAdapatation
+        self.paraLimit = 16
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: same thing, and can you clarify what `para_limit` means?

File: nevergrad/optimization/optimizerlib.py [hunk 3]
```diff
@@ -1522,35 +1531,63 @@ def _internal_provide_recommendation(self) -> ArrayLike:
...
         candidate = parent.spawn_child().set_standardized_data(individual, reference=self.parametrization)
         if parent is self.parametrization:
             candidate.heritage["lineage"] = candidate.uid
-        candidate._meta["sigma"] = self.sigma
+        candidate._meta["sigma"] = tmp
         return candidate
 
     def _internal_tell_candidate(self, candidate: p.Parameter, value: float) -> None:
         candidate._meta["loss"] = value
+        if self.populationSizeAdapatation:
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: to simplify the code you can probably always use the mu and llambda from the popsize instance, and either update it with add_value, or manually by directly overriding llambda and mu. This way there would be a unique source of truth for llambda and mu

File: nevergrad/optimization/optimizerlib.py [hunk 4]
```diff
@@ -1522,35 +1531,63 @@ def _internal_provide_recommendation(self) -> ArrayLike:
...
-                    for i in range(self.mu)]
-            if self.isotropic:
-                self.sigma = np.sqrt(sum(stdd) / (self.mu * self.dimension))
+                                      for c in self.parents) / mu
+            if self.populationSizeAdapatation:
+                if self.popsize.llambda < self.paraLimit * self.dimension: # Population size not large enough for emna
+                    if self.isotropic:
+                        self.sigma = np.exp(np.sum(np.log([c._meta["sigma"] for c in self.parents])) / mu)
+                    else:
+                        self.sigma = np.exp(np.sum(np.log([c._meta["sigma"] for c in self.parents]), axis=0) / mu)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: I think you can combine both lines to 
```python
self.sigma = np.exp(np.sum(np.log([c._meta["sigma"] for c in self.parents]), axis=0 if self.isotropic else None) / mu)
```

File: nevergrad/optimization/optimizerlib.py [hunk 5]
```diff
@@ -1522,35 +1531,63 @@ def _internal_provide_recommendation(self) -> ArrayLike:
...
+                else:
+                    stdd = [(self.parents[i].get_standardized_data(reference=self.parametrization) - self.current_center)**2 for i in range(mu)]
+                    if self.isotropic:
+                        self.sigma = np.sqrt(sum(stdd) / (mu * self.dimension))
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: same as above, and to share as much code as possible then maybe just use a `axis` variable:
`axis = 0 if self.isotropic else None`

File: nevergrad/optimization/optimizerlib.py [hunk 6]
```diff
@@ -1522,35 +1531,63 @@ def _internal_provide_recommendation(self) -> ArrayLike:
...
     def _internal_tell_candidate(self, candidate: p.Parameter, value: float) -> None:
         candidate._meta["loss"] = value
+        if self.populationSizeAdapatation:
+            self.popsize.add_value(value)
         self.children.append(candidate)
-        if len(self.children) >= self.llambda:
+        popSize = -1
+        mu = -1
+        if self.populationSizeAdapatation:
+            popSize = self.popsize.llambda
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: please you pep8 compatible naming as well here

**Comment by fteytaud** [comment]
Thx for all your remarks ; it is still in progress, the code is dirty. I PR it, just to test something with pytest/mypy), I didn't want to bother you.

**Comment by fteytaud** [comment]
Ok I felt guilty that you read my ugly code so I have cleaned it ; I think it may be ok now :-)

File: nevergrad/optimization/optimizerlib.py [hunk 7]
```diff
@@ -1492,24 +1492,32 @@ def __init__(
...
+        # population size and parent size initializations
+        dim = self.dimension
+        self.popsize = _PopulationSizeController(llambda=4 * dim, mu=dim, dimension=dim, num_workers=num_workers)
+        if not self.population_size_adaptation:
+            self.popsize.mu = max(16, self.dimension)
+            self.popsize.llambda = 4 * self.popsize.mu
+            if budget is not None and self.popsize.llambda > budget:
+                self.popsize.llambda = budget
+                self.popsize.mu = self.popsize.llambda // 4
+                warnings.warn("Budget may be too small in front of the dimension for EMNA")
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: can you send a base.InefficientSettingsWarning instead?

https://github.com/facebookresearch/nevergrad/blob/b01c503748ed2dbbe0c3e0bc4f49baf2964c458a/nevergrad/optimization/differentialevolution.py#L87-L88

This makes the warning uniform within nevergrad, and I filter them out during unit tests

File: nevergrad/optimization/optimizerlib.py [hunk 8]
```diff
@@ -1522,35 +1530,46 @@ def _internal_provide_recommendation(self) -> ArrayLike:
             return self.current_center
 
     def _internal_ask_candidate(self) -> p.Parameter:
-        individual = self.current_center + self.sigma * self._rng.normal(0, 1, self.dimension)
+        sigma_tmp = self.sigma
+        if self.population_size_adaptation and self.popsize.llambda < self.min_coef_parallel_context * self.dimension:
+            mutated_sigma = self.sigma * np.exp(self._rng.normal(0, 1) / np.sqrt(self.dimension))
+            individual = self.current_center + mutated_sigma * self._rng.normal(0, 1, self.dimension)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: I think you can move this line out and use sigma_tmp instead?

**Comment by fteytaud** [comment]
Thx for your remarks. I think that's ok ; I think we can improve the algoritlm with a better tuning, I will do that later.

**Pull request merged**